File: WebAPI开发手册/demo/cloud_http_ocr_test.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

"""
python 3.7
pip install requests
"""

import base64
import hashlib
import json
import time
import uuid
from email.utils import formatdate

import requests

# 由北京市大牛儿科技发展有限公司统一分配
org_id = 0
app_id = 0
app_key = ""

# 图片上文本的语种
# chs： 简体中文
# cht： 繁体中文
language = "chs"

# 请求的地址
request_url = "http://api.baller-tech.com/v1/service/v1/ocr"
# 测试使用的图像数据
image_file = ""
# 推送结果的地址，该地址为调用者自己搭建的接收推送结果的Web服务地址
callback_url = ""


def post_data(request_id, data):
    """
    向服务器端发送图像数据
    :param request_id: 请求ID
    :param data: 图像数据
    :return: True：成功；False：失败
    """
    business_params = {
        'request_id': str(request_id),
        'image_mode': 'multi_row',
        "language": language,
    }
    if callback_url:
        business_params["callback_url"] = callback_url
    business_params_base64 = base64.b64encode(json.dumps(business_params).encode(encoding='utf-8')).decode()

    request_time = formatdate(timeval=None, localtime=False, usegmt=True)
    before_md5 = app_key + request_time + business_params_base64
    hl = hashlib.md5()
    hl.update(before_md5.encode(encoding='utf-8'))
    headers = {
        'B-AppId': str(app_id),
        'B-CurTime': request_time,
        'B-Param': business_params_base64,
        'B-CheckSum': hl.hexdigest(),
        'Content-Type': "application/octet-stream"
    }

    response = requests.post(request_url, headers=headers, data=data)
    try:
        response_content = json.loads(response.content)
    except Exception as e:
        print(f"{request_id} load POST response {response.content} failed {e}")
        return False

    if 0 != response_content["code"]:
        print(f"{request_id} POST failed {response_content['code']} {response_content['message']} ")
        return False

    return True


def get_result(request_id):
    """
    获取识别结果
    :param request_id: 请求ID
    :return: True：结果获取结束；False：结果获取未结束
    """
    business_params = {
        'request_id': request_id,
    }
    business_params_base64 = base64.b64encode(json.dumps(business_params).encode(encoding='utf-8')).decode()

    request_time = formatdate(timeval=None, localtime=False, usegmt=True)
    before_md5 = app_key + request_time + business_params_base64
    hl = hashlib.md5()
    hl.update(before_md5.encode(encoding='utf-8'))
    headers = {
        'B-AppId': str(app_id),
        'B-CurTime': request_time,
        'B-Param': business_params_base64,
        'B-CheckSum': hl.hexdigest(),
    }

    response = requests.get(request_url, headers=headers)
    try:
        response_content = json.loads(response.content)
    except Exception as e:
        print(f"{request_id} load GET response {response.content} failed {e}")
        return True

    recognition_result = {}
    if 0 == response_content["code"]:
        # 图像识别时，会将图片中的信息按照一定的规则（目前按行）分为不同的子项，每个GET获取的是一个或多个子项的结果
        for sub_item_result in response_content["data"]:
            if sub_item_result["result"]:
                recognition_result[sub_item_result['order']] = sub_item_result['result']

        if recognition_result:
            sorted(recognition_result.keys())
            for _, value in recognition_result.items():
                print(value)
    else:
        print(f"{request_id} GET failed {response_content['code']} {response_content['message']}")
    return 1 == response_content["is_end"]


def main():
    # 读取图像数据
    with open(image_file, "rb") as file:
        image_data = file.read()
    if not image_data:
        print(f"read image file {image_file} failed")
        return

    # 发送图像数据
    request_id = str(uuid.uuid1())
    put_success = post_data(request_id, image_data)
    if not put_success:
        print(f"{request_id} POST data failed")
        return

    # 获取识别结果
    if not callback_url:
        is_end = False
        while not is_end:
            is_end = get_result(request_id)
            time.sleep(0.150)
        print(f"{request_id} GET result finished")


if __name__ == '__main__':
    main()

File: WebAPI开发手册/demo/cloud_http_ocr_test.java
package com.baller;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;
import java.util.UUID;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.util.DigestUtils;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

public class cloud_http_ocr_test {
  
  //地址配置
  public static String mUrl = "http://api.baller-tech.com/v1/service/v1/ocr";

  // 账号信息 由北京市大牛儿科技发展有限公司统一分配;
  public static long mAppId = 0L;
  public static String mAppkey = "";

  // 测试语种
  // 请查考《图像识别（OCR）HTTP协议WebAPI开发文档.pdf》中“支持的语种”章节
  public static String mLanguage = "chs";
  // 测试使用的文件
  public static String mTestFile = "";
  // 请查考《图像识别（OCR）HTTP协议WebAPI开发文档.pdf》中“PDF识别注意事项”章节
  public static String mFileFormat = "";
  public static String mImageMode = "multi_row";
  // 结果保存文件
  public static String mSaveResultFile = mTestFile + "_out.txt";
  // 推送结果的地址，该地址为调用者自己搭建的接收推送结果的Web服务地址
  public static String mCallbackUrl = "";
  // 结果是否保存到文件中
  public static boolean mSaveToFile = true;

  @SuppressWarnings("resource")
  public static byte[] readFile(String strFile) {
    byte[] fileData = null;
    BufferedInputStream inputStream;
    try {
      inputStream = new BufferedInputStream(new FileInputStream(strFile));
    } catch (FileNotFoundException e) {
      e.printStackTrace();
      return fileData;
    }

    try {
      fileData = new byte[inputStream.available()];
    } catch (IOException e) {
      e.printStackTrace();
      return fileData;
    }

    try {
      inputStream.read(fileData);
    } catch (IOException e) {
      e.printStackTrace();
      return fileData;
    }

    return fileData;
  }

  public static String getGmtTime() {
    SimpleDateFormat sdf3 = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US);
    sdf3.setTimeZone(TimeZone.getTimeZone("GMT"));
    return sdf3.format(new Date());
  }

  @SuppressWarnings({ "unchecked", "rawtypes" })
  public static boolean postData(String requestId, byte[] testData) {

    Map<String, String> businessParams = new HashMap<>();
    businessParams.put("request_id", requestId);
    businessParams.put("language", mLanguage);
    businessParams.put("image_mode", mImageMode);
    businessParams.put("file_format", mFileFormat);
    businessParams.put("input_mode", "once");
    if (!mCallbackUrl.isEmpty()) {
      businessParams.put("callback_url", mCallbackUrl);
    }
    String businessParamsBase64 = Base64.getEncoder().encodeToString(
        JSON.toJSONString(businessParams).getBytes());

    String requestTime = getGmtTime();
    String checkSum = mAppkey + requestTime + businessParamsBase64;
    String md5 = DigestUtils.md5DigestAsHex(checkSum.getBytes());

    MultiValueMap headers = new LinkedMultiValueMap<String, String>();
    headers.set("B-AppId", String.valueOf(mAppId));
    headers.set("B-CurTime", requestTime);
    headers.set("B-Param", businessParamsBase64);     
    headers.set("B-CheckSum", md5);
    headers.set("Content-Type", "application/octet-stream");

    HttpEntity<String> param = new HttpEntity(testData, headers);
    RestTemplate restTemplate = new RestTemplate();
    ResponseEntity<String> response = restTemplate.exchange(
        mUrl, HttpMethod.POST, param, String.class);
    JSONObject responseContent = JSONObject.parseObject(response.getBody());

    if (responseContent.getInteger("code") != 0) {
      System.out.println(requestId + " POST failed. " + responseContent.getInteger("code") + " " 
            +  responseContent.get("message"));
      return false;
    }
    return true;
  }

  @SuppressWarnings({ "rawtypes", "unchecked"})
  public static boolean getResult(String requestId, FileOutputStream outFile) {
 
    Map<String, String> paramMap = new HashMap<>();
    paramMap.put("request_id", requestId);
    String businessParamBase64 = Base64.getEncoder().encodeToString(
        JSON.toJSONString(paramMap).getBytes());

    String requestTime = getGmtTime();
    String checkSum = mAppkey + requestTime + businessParamBase64;
    String md5 = DigestUtils.md5DigestAsHex(checkSum.getBytes());

    MultiValueMap headerMap = new LinkedMultiValueMap<String, String>();
    headerMap.set("B-AppId", String.valueOf(mAppId));
    headerMap.set("B-CurTime", requestTime);
    headerMap.set("B-Param", businessParamBase64);     
    headerMap.set("B-CheckSum", md5);

    HttpEntity<String> param = new HttpEntity(headerMap);
    RestTemplate restTemplate = new RestTemplate();
    ResponseEntity<String> exchange = restTemplate.exchange(
        mUrl, HttpMethod.GET, param, String.class);

    JSONObject jsonObject = JSONObject.parseObject(exchange.getBody());

    if (jsonObject.getInteger("code") != 0) {
      System.out.println(requestId + " GET failed. " + jsonObject.getInteger("code") + " "
            +  jsonObject.get("message"));
      return true;
    }

    JSONArray dataArray = jsonObject.getJSONArray("data");
    if (!dataArray.isEmpty()) {
      ArrayList<String> dataList = new ArrayList<String>();

      for (int i = 0; i < dataArray.size(); i++) {
        JSONObject dataJson = dataArray.getJSONObject(i);
        if (!dataJson.getString("result").isEmpty()) {
          dataList.add(dataJson.getIntValue("order"), dataJson.getString("result"));
        }
      }

      if (!dataList.isEmpty()) {
        for (int j = 0; j < dataList.size(); j++) {
          System.out.println(dataList.get(j));
          if (outFile != null) {
            try {
              outFile.write(dataList.get(j).getBytes());
              String changeLine = "\n";
              outFile.write(changeLine.getBytes());
            } catch (IOException e) {
              // TODO Auto-generated catch block
              e.printStackTrace();
            }
          }
        }
      }
    }
    return jsonObject.getInteger("is_end") == 1;
  }

  /**
   * once 方式发送.
   */
  public static void testOnce() {
    String requestId = UUID.randomUUID().toString();

    FileOutputStream fileOut = null;
    if (mSaveToFile) {
      try {
        File f = new File(mSaveResultFile);
        fileOut = new FileOutputStream(f);
      } catch (FileNotFoundException e1) {
        e1.printStackTrace();
      }
    }

    // 读取并发送图像数据
    boolean putSuccess = postData(requestId, readFile(mTestFile));
    if (!putSuccess) {
      System.out.println(requestId + " POST data failed");
      return;
    }

    if (mCallbackUrl.isEmpty()) {
      // 获取识别结果
      while (!getResult(requestId, fileOut)) {
        try {
          // 停留40ms的作用：
          // 1. 避免线程一直被占用，导致CPU利用率高。40ms为经验值，使用时可根据实际的需求调大或调小。
          Thread.sleep(40);
        } catch (InterruptedException e) {
          // TODO Auto-generated catch block
          e.printStackTrace();
        }
      }
    }

    if (fileOut != null) {
      try {
        fileOut.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
    System.out.println(requestId + " GET result finished");
  }

  public static void main(String[] args) {
    testOnce();
  }
}